### ISSUE TITLE
Type of variable isn't widened inside a conditional inside a loop

### ISSUE THREAD

**Issue by icecream17** [microsoft/TypeScript#46475]
# Bug Report

### 🔎 Search Terms

ts2367 widen loop

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

[Playground link](https://www.typescriptlang.org/play?#code/FAGwpgLgBAxg9gEzALigBigHygRi1AJigF51h4A7AZ2jggAswAnAYUTCtQAoNs9sCASgDaAXRJRhBADTpZOWWlkzCyxfPWrcGuVpyjgwAGZwmULpRqx2ROEah1GrdlUFQA3sCjeoAS3sWNiTEpGhu7j6RUF5R3vBIJDGxkfFgwaQEScmxAPy4WdmRqGhZAL5QYCBUaZ7ZqRKZkaXAzUA)

### 💻 Code

__Edit: shorter version by gabritto__

```ts
let code: 0 | 1 = 0
while (true) {
    code = code === 1 ? 0 : 1
}
```

And in practice:
![image](https://user-images.githubusercontent.com/58114641/138378445-b5ec3475-d24c-45e8-ac34-953d10893106.png)

### 🙁 Actual behavior

Error

### 🙂 Expected behavior

Since `code` could be changed in previous iterations of the loop, the condition is not always false, so there shouldn't be any error.

**Comment by MartinJohns** [comment]
Your code makes no sense. You're checking if code equals 0, and if it does you want to compare if with 2. It can never equal 2, because you made sure it equals 0 the line before.

**Comment by andrewbranch** [comment]
A little more reduced: https://www.typescriptlang.org/play?ts=4.6.0-dev.20211105#code/FAGwpgLgBAxg9gEzALigBigHygRi1AJigF51hgAzOAJygAoBuBgSigG9govZEwSekJYqSIB+XFFRpy3KAEsK9AIRKKAQxABnMKw6zZ8QSM7cAvsHNA

The code in the screenshot is kind of easier to follow even though it’s longer since it’s real.

**Comment by jameswilddev** [comment]
Question: I've just encountered this which looks similar to me, the root problem in both cases being that TypeScript is very quick to narrow types on `let` statements when it should probably leave them as the type they were declared as:

https://www.typescriptlang.org/play?#code/DYUwLgBGAWCWB2BzAXBABgQTRAPugQtnmgMLYC86WA3ALABQDAxgPbwDOk74ElAFGBapMRAqNJoAlKgBuLWABNeAPggBvBhC1Q4SXlBZ16AXyMNuYPhMln6sAGYQBuxL3KVr6zdtYcWoADpgFkQ+AHJEFjCbBmMGBiA

Would you say this is the same issue or that I should raise a seperate bug?

**Comment by Zzzen** [comment]
@jameswilddev it's duplicate of https://github.com/microsoft/TypeScript/issues/44824

**Comment by gabritto** [comment]
Even smaller: https://www.typescriptlang.org/play?ts=4.5.4#code/DYUwLgBAxg9gJiAXBADBAPhAjBAvKgKAHcALAS1AgAowAnAVxAEoIBvAiT6eEPDrgQNgI8ufFn6CpXAPyFpCiMiwBuAgF8gA

**Comment by MartinJohns** [comment]
@RyanCavanaugh Probably want to update the milestone again. :-)

**Comment by gabritto** [comment]
@ahejlsberg I started investigating this bug a while back but couldn't come up with a fix, so it was reassigned to you, but now I'm wondering if I could fix this along with a fix for #47539. Are you planning on working on this?

**Comment by ahejlsberg** [comment]
The issue here is that during control flow analysis of the loop, the type of the `code` variable is seen as `0` during an intermediate step. That's to be expected, but unfortunately it triggers an error that isn't actually correct once we arrive at the full type `0 | 1`. We need to suppress that error, which we can do by tracking the fact that `checkExpression` was called from `getTypeOfExpression` (in which we care only about the type and not the diagnostics).
